### PR TITLE
[OCPCLOUD-994] Add simple overview user document

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Describe which components are located under MAO management, how the operator components are deployed by the CVO, and the ClusterOperator status conditions.

### CODE REVIEW

File: docs/user/Overview.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+## CVO integration
+
+`MAO` code is running under the `machine-api-operator` `Deployment`. This and other resources listed below are relying on `CVO` management strategy, and are being consumed from the `/install` directory: https://github.com/openshift/machine-api-operator/tree/master/install
+
+Namely:
+- `openshift-machine-api` `Namespace` - the primary and the only namespace, where the `MAO` is hosted, and all of the dependent resources are being managed by the operator.
+- `machine-api-operator` `Deployment` - `MAO` code runs here.
+- `machine-api-operator-images` `ConfigMap` - is injected with provider images, used by `MAO` to deploy a cloud-provider specific code.
+- `CredentialsRequests` - granular request for cloud credentials, served by https://github.com/openshift/cloud-credential-operator
```
Comment (JoelSpeed) [inline docs/user/Overview.md:13]:
```suggestion
- `CredentialsRequests` - granular request for cloud credentials, fulfilled by the [cloud-credential-operator](https://github.com/openshift/cloud-credential-operator)
```

File: docs/user/Overview.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+- `machine-api-operator` `Deployment` - `MAO` code runs here.
+- `machine-api-operator-images` `ConfigMap` - is injected with provider images, used by `MAO` to deploy a cloud-provider specific code.
+- `CredentialsRequests` - granular request for cloud credentials, served by https://github.com/openshift/cloud-credential-operator 
+- `Alerts`, `Metrics`, `CRDs`, `RBAC`, and `Services` are deployed from the `/install` directory.
+
+### Upgrade
+
+`MAO` installation/upgrade procedure relies on `CVO`. 
+
+`MAO` is deployed after `CVO` starts running at [run level](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level) 30, right after the Kubernetes operators (run level 10-29), and before the machine-config-operator (run level 80)
```
Comment (JoelSpeed) [inline docs/user/Overview.md:20]: Good link there! Would have asked for that if you hadn't put it in! 😄

File: docs/user/Overview.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+`MAO` hosts `CustomResourceDefinitions` and controllers, responsible for serving each of the following resources.
+- `Machines`
+- `MachineSets`
+- `MachineHealthChecks`
+
+This operator is responsible for the creation and maintenance of:
+- `machine-api-operator` `ClusterOperator` - `MAO` status reporting
+- `machine-api-controllers` `Deployment` - controllers for all supported `CRDs`
+- `machine-api` `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` - validation and defaulting for Machine resources
+termination handler `DaemonSet` - monitoring for spot instances state and remediating `Machines`, which are deployed on those in case the instance goes away.
```
Comment (JoelSpeed) [inline docs/user/Overview.md:37]: Think this should be a bullet point as well right?

File: docs/user/Overview.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+### Implementing
+
+- `Machine` controller - manages Machine resources. It uses actuator interface, which follows a `Machine` lifecycle [pattern](https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md) This interface provides `Create`, `Update`, and `Delete` methods to manage your provider specific cloud instances, connected storage, and networking settings to make the instance prepared for bootstrapping. Each provider is therefore responsible for implementing these methods.
```
Comment (JoelSpeed) [inline docs/user/Overview.md:41]: Maybe link `actuator interface` to the code?

File: docs/user/Overview.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+- [Azure](https://github.com/openshift/cluster-api-provider-azure)
+- [BareMetal](https://github.com/openshift/cluster-api-provider-baremetal/)
+- [OVirt](https://github.com/openshift/cluster-api-provider-ovirt)
+
+## Works closely, but not directly responsible for
+
+- [cluster-api-actuator-pkg](https://github.com/openshift/cluster-api-actuator-pkg/) - hosting e2e tests for `MAO` and supported cloud providers.
+machine-config-operator - creating machineConfigs with configuration to inject into provisioned Machine instances. Is responsible for initiating Node bootstrap procedure for newly created Machine.
+- [cluster-machine-approver](https://github.com/openshift/cluster-machine-approver) - approving Node CSR for newly provisioned Machine.
+- [cluster-autoscaler](https://github.com/openshift/kubernetes-autoscaler/tree/master/cluster-autoscaler) - automatic scaling of MachineSet resources, manages ClusterAutoscaler and MachineAutoscaler resources.
```
Comment (JoelSpeed) [inline docs/user/Overview.md:62]: > manages ClusterAutoscaler and MachineAutoscaler resources.

This is cluster-autoscaler-operator not cluster-autoscaler, should include that repo linked here as well

File: docs/user/Overview.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+## ClusterOperator
+
+### Status management
+
+`MAO` is responsible for status reporting on the `machine-api` `ClusterOperator`. Our status reporting  is following the [best practices](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#conditions).
+
+The status condition goes `Degraded` if any of the managed resources are failed the rollout procedure, or were unavailable for longer [periods](https://github.com/openshift/machine-api-operator/blob/760561cc29c823a62e2f4c9d7798592d3107e45d/pkg/operator/sync.go#L31-L34) of time.
```
Comment (JoelSpeed) [inline docs/user/Overview.md:70]: I think this improves readability
```suggestion
The status condition will turn `Degraded` if any of the managed resources fail to rollout, or are unavailable for longer [periods](https://github.com/openshift/machine-api-operator/blob/760561cc29c823a62e2f4c9d7798592d3107e45d/pkg/operator/sync.go#L31-L34) of time. 
```

**Review by JoelSpeed — COMMENTED** [review]
Please link the relevant Jira card in the PR title. Added some suggestions for improvements to the readabilty primarily, content is good otherwise

Not sure whether we want to use a more specific name for the file here? I suspect we will have overviews of a number of things in this folder so maybe `machine-api-operator-overview.md` would be better?

File: docs/user/machine-api-operator-overview.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+This document describes how the  Machine API Operator (herein `MAO`) gets deployed, and the `MAO`'s responsibilities. It includes guidelines on how to troubleshoot typical issues with resources provisioned by `MAO`.
+
+## CVO integration
+
+`MAO` code is running under the `machine-api-operator` `Deployment`. This and other resources listed below rely on the `CVO` management strategy. These resources are consumed from the [`/install` directory]( https://github.com/openshift/machine-api-operator/tree/master/install).
+
+Namely:
+- `openshift-machine-api` `Namespace` - the primary and the only namespace, where the `MAO` is hosted, and all of the dependent resources are being managed by the operator.
+- `machine-api-operator` `Deployment` - `MAO` code runs here.
+- `machine-api-operator-images` `ConfigMap` - is injected with provider images, used by `MAO` to deploy cloud-provider specific code.
```
Comment (michaelgugino) [inline docs/user/machine-api-operator-overview.md:12]: TODO: add a section with a brief explanation about how this data is injected and who generates those images (eg, release tooling).  Too much info for this line, doesn't need to be this patch set.

Comment (Danil-Grigorev) [reply docs/user/machine-api-operator-overview.md:12]: I'll add a comment in the doc, so we could return to this question later.

File: docs/user/machine-api-operator-overview.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+`MAO` is responsible for status reporting on the `machine-api` `ClusterOperator`. Our status reporting  is following the [best practices](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#conditions).
+
+The status condition will turn `Degraded` if any of the managed resources fail to rollout, or are unavailable for longer [periods](https://github.com/openshift/machine-api-operator/blob/760561cc29c823a62e2f4c9d7798592d3107e45d/pkg/operator/sync.go#L31-L34) of time. 
+
+In addition to the cluster-operator status reporting, it is recommended to know relevant alerts described in the alerting [document](https://github.com/openshift/machine-api-operator/blob/master/docs/user/Alerts.md)
+
+## Troubleshooting
+
+Q: "any" of the components maintained by `MAO` is lagging/missing/unchanged for a long period of time.
```
Comment (elmiko) [inline docs/user/machine-api-operator-overview.md:77]: no need for quotes on "any" here

File: docs/user/machine-api-operator-overview.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+A: check the status of the machine-api-operator deployment, if it is running all the replicas. Check the machine-api-operator logs. Check if the machine-api cluster-operator didn’t go `Degraded`, or the `MachineApiOperatorDown` alert was not firing.
```
Comment (elmiko) [inline docs/user/machine-api-operator-overview.md:78]: i would add the code blocks around "machine-api-operator" here

File: docs/user/machine-api-operator-overview.md (hunk 7 above)
Comment (Danil-Grigorev) [reply docs/user/machine-api-operator-overview.md:12]: @michaelgugino  Realized the comment is not what was needed, so there is a breif overview of `ImageStreams` we use.

**Review by elmiko — COMMENTED** [review]
@Danil-Grigorev i think you may have misinterpreted my comment about `MAO`. i think we should just say MAO instead of `MAO` in the code. 

same thing with CVO, and other abbreviations for operators. if you want to highlight them i would use **MAO** bold instead of code blocks.

**Review by elmiko — COMMENTED** [review]
Danil, the content here looks good to me. i would prefer if we didn't use code blocks for all the resource types, for example i don't think we should say "the machine-api-operator `Deployment`". there are a couple places i think we could use the code blocks, but in general i prefer to reserve their use for things that should be typed into a terminal or console, and literal code pieces.

i think we should strive to follow the openshift docs guidlines when we can, https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc

File: docs/user/machine-api-operator-overview.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+- ImageStreams for the ConfigMap with provider image mapping - a set of plain DockerImages, injected into OCP release payload.
+- Alerts, Metrics, CRDs, RBAC, and Services are deployed from the "/install" directory.
+
+### Upgrade
+
+MAO installation/upgrade procedure relies on CVO. 
+
+MAO is deployed after CVO starts running at [run level](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level) 30, right after the Kubernetes operators (run level 10-29), and before the `machine-config-operator` (run level 80)
+
+The new version of MAO Deployment is responsible for installing/upgrading provider executables
```
Comment (elmiko) [inline docs/user/machine-api-operator-overview.md:23]: the "new version" part here is a little confusing to me, what is this referring to?

File: docs/user/machine-api-operator-overview.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+- [cluster-machine-approver](https://github.com/openshift/cluster-machine-approver) - approving Node CSR for newly provisioned Machine.
+- [cluster-autoscaler-operator](https://github.com/openshift/cluster-autoscaler-operator) - automatic scaling of MachineSet resources, manages ClusterAutoscaler and `MachineAutoscaler` resources.
+- [release](https://github.com/openshift/release) - the tooling responsible for building openshift components and images, including MAO.
+- [installer](https://github.com/openshift/installer) - provision the initial cluster infrastructure (`IPI`) from a scratch, which is later used by MAO to manipulate `VMs`, network and storage configuration for worker Machines.
+
+## ClusterOperator
+
+### Status management
+
+MAO is responsible for status reporting on the `machine-api` ClusterOperator. Our status reporting  is following the [best practices](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#conditions).
```
Comment (elmiko) [inline docs/user/machine-api-operator-overview.md:71]: minor nit
```suggestion
MAO is responsible for status reporting on the `machine-api` ClusterOperator. Our status reporting is following the OpenShift [best practices](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#conditions).
```

File: docs/user/machine-api-operator-overview.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+## Troubleshooting
+
+Q: Any of the components maintained by MAO is lagging/missing/unchanged for a long period of time.
+A: check the status of the `machine-api-operator` deployment, if it is running all the replicas. Check the `machine-api-operator` logs. Check if the `machine-api` cluster-operator didn’t go `Degraded`, or the `MachineApiOperatorDown` alert was not firing. 
+
+Q: MAO deployment is outdated/missing
+A: check the CVO health by checking ClusterVersion object ([guide](https://github.com/openshift/cluster-version-operator/blob/master/docs/user/status.md)) It should be `Available` and not `Progressing`.
+
+More detailes on troubleshooting you will find in the troubleshooting [guide](https://github.com/openshift/machine-api-operator/blob/master/docs/user/TroubleShooting.md).
```
Comment (elmiko) [inline docs/user/machine-api-operator-overview.md:85]: minor nit

s/detailes/details/

**Review by elmiko — COMMENTED** [review]
looking good Danil, i just have a question and a couple minor suggestions

File: docs/user/machine-api-operator-overview.md (hunk 10 above)
Comment (Danil-Grigorev) [reply docs/user/machine-api-operator-overview.md]: What I wanted to say "rolled-out version of MAO Deployment is ..." as trying to describe upgrade order or procedure. Good question.

Comment (elmiko) [reply docs/user/machine-api-operator-overview.md]: maybe could say "A newly deployed version of MAO is responsible for ...."?

File: docs/user/machine-api-operator-overview.md (hunk 11 above)
Comment (elmiko) [reply docs/user/machine-api-operator-overview.md:71]: did this get missed or ?

File: docs/user/machine-api-operator-overview.md (hunk 10 above)
Comment (Danil-Grigorev) [reply docs/user/machine-api-operator-overview.md]: Done

**Comment by Danil-Grigorev** [comment]
> Danil, the content here looks good to me. i would prefer if we didn't use code blocks for all the resource types, for example i don't think we should say "the machine-api-operator `Deployment`". there are a couple places i think we could use the code blocks, but in general i prefer to reserve their use for things that should be typed into a terminal or console, and literal code pieces or exact values a user should watch for.
> 
> i think we should strive to follow the openshift docs guidlines when we can, https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc

That's actually correct, as it described in the doc - https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#object-references

**Review by elmiko — APPROVED** [review]
i'm good with this, thanks Danil!
/approve

**Comment by elmiko** [comment]
> That's actually correct, as it described in the doc - https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#object-references

i was looking at the section before that one, but i think it's fine the way it is.

**Review by alexander-demichev — APPROVED** [review]
/lgtm

**Pull request merged**